File: project/src/manager/orchestrator.py
"""Main Multi-Agent Orchestrator engine."""

import os
import time
from typing import Any, Dict, List, Optional

from project.src.common.config import load_config_file
from project.src.common.constants import DEFAULT_MAX_RETRY, DEFAULT_WORKER_PORTS
from project.src.common.schemas import (
    AgentRole,
    AgentStatus,
    TaskItem,
    TaskRequest,
    TaskResult,
    TaskStatus,
)
from project.src.manager.db import DatabaseManager
from project.src.manager.evaluator import EvaluationVerdict, ResultEvaluator
from project.src.manager.git_manager import GitManager
from project.src.manager.task_manager import TaskManager
from project.src.manager.worker_client import WorkerClient


class Orchestrator:
    def __init__(
        self,
        config_path: str = "project/project.yaml",
        db_path: str = "logs/database.sqlite",
        repo_path: str = ".",
    ):
        self.config = load_config_file(config_path)
        self.db = DatabaseManager(db_path=db_path)
        self.task_manager = TaskManager(db=self.db)
        self.evaluator = ResultEvaluator(max_retries=self.config.get("manager", {}).get("max_retries", DEFAULT_MAX_RETRY))
        self.git_manager = GitManager(repo_path=repo_path)
        self.worker_client = WorkerClient()
        self.repo_path = os.path.abspath(repo_path)

        # Worker agent network mappings
        self.agents_config = self.config.get("agents", {})

    def get_agent_endpoint(self, agent_id_or_role: str) -> Dict[str, Any]:
        """Find host and port for an agent ID or role."""
        for role_name, info in self.agents_config.items():
            if isinstance(info, dict):
                if info.get("id") == agent_id_or_role or role_name == agent_id_or_role or info.get("role") == agent_id_or_role:
                    return {
                        "id": info.get("id", f"agent-{role_name}"),
                        "host": info.get("host", "127.0.0.1"),
                        "port": info.get("port", DEFAULT_WORKER_PORTS.get(info.get("id", ""), 8001)),
                        "role": info.get("role", role_name),
                    }

        default_port = DEFAULT_WORKER_PORTS.get(agent_id_or_role, 8001)
        return {
            "id": agent_id_or_role,
            "host": "127.0.0.1",
            "port": default_port,
            "role": "developer",
        }

    def check_all_agents_health(self) -> Dict[str, bool]:
        health_status = {}
        for role_name, info in self.agents_config.items():
            if isinstance(info, dict):
                agent_id = info.get("id", role_name)
                host = info.get("host", "127.0.0.1")
                port = info.get("port", 8001)
                is_healthy = self.worker_client.check_health(host, port)
                health_status[agent_id] = is_healthy
        return health_status

    def decompose_requirement(self, prompt: str, target_lang: str = "python") -> List[TaskItem]:
        """Decompose a high-level requirement into a standard task DAG:

        1. Research/Doc -> 2. Implementation -> 3. Testing
        -> 4. Review -> 5. Security
        """
        all_existing = self.task_manager.list_tasks()
        base_num = len(all_existing) + 1

        id1 = f"TASK-{base_num:03d}"
        id2 = f"TASK-{base_num + 1:03d}"
        id3 = f"TASK-{base_num + 2:03d}"
        id4 = f"TASK-{base_num + 3:03d}"
        id5 = f"TASK-{base_num + 4:03d}"

        tasks = []

        # 1. Research & Spec Task
        t1 = self.task_manager.add_task(
            task_id=id1,
            title=f"Spec & Architecture: {prompt[:50]}",
            description=f"Analyze requirements and produce architecture notes for: {prompt}",
            assigned_agent="agent-e",
            role=AgentRole.RESEARCHER,
            priority="high",
            dependencies=[],
        )
        tasks.append(t1)

        # 2. Developer Implementation Task
        t2 = self.task_manager.add_task(
            task_id=id2,
            title=f"Implementation: {prompt[:50]}",
            description=f"Implement source code for: {prompt}",
            assigned_agent="agent-a",
            role=AgentRole.DEVELOPER,
            priority="high",
            dependencies=[id1],
        )
        tasks.append(t2)

        # 3. Tester Verification Task
        t3 = self.task_manager.add_task(
            task_id=id3,
            title=f"Testing & Build: {prompt[:50]}",
            description=f"Run test suite and verify build for implementation of: {prompt}",
            assigned_agent="agent-b",
            role=AgentRole.TESTER,
            priority="high",
            dependencies=[id2],
        )
        tasks.append(t3)

        # 4. Reviewer Code Review Task
        t4 = self.task_manager.add_task(
            task_id=id4,
            title=f"Code Review: {prompt[:50]}",
            description=f"Perform static analysis and review code quality for: {prompt}",
            assigned_agent="agent-c",
            role=AgentRole.REVIEWER,
            priority="medium",
            dependencies=[id3],
        )
        tasks.append(t4)

        # 5. Security Scan Task
        t5 = self.task_manager.add_task(
            task_id=id5,
            title=f"Security Audit: {prompt[:50]}",
            description=f"Audit vulnerabilities, secret leaks, and security posture for: {prompt}",
            assigned_agent="agent-d",
            role=AgentRole.SECURITY,
            priority="medium",
            dependencies=[id4],
        )
        tasks.append(t5)

        return tasks

    def execute_single_task(self, task: TaskItem, metadata: Optional[Dict[str, Any]] = None) -> TaskResult:
        endpoint = self.get_agent_endpoint(task.assigned_agent)
        host = endpoint["host"]
        port = endpoint["port"]

        self.task_manager.update_task_status(task.task_id, TaskStatus.RUNNING)

        task_req = TaskRequest(
            task_id=task.task_id,
            type=task.role.value if hasattr(task.role, "value") else str(task.role),
            title=task.title,
            description=task.description,
            repository=self.repo_path,
            branch=f"{task.assigned_agent}/{task.task_id.lower()}",
            timeout_seconds=300,
            metadata=metadata or {},
        )

        # Send task
        send_res = self.worker_client.send_task(host, port, task_req)
        if "error" in send_res:
            # Worker not reachable or error
            err_result = TaskResult(
                task_id=task.task_id,
                agent_id=task.assigned_agent,
                status="FAILED",
                summary=f"Failed to communicate with worker {task.assigned_agent} on {host}:{port}: {send_res.get('error')}",
                errors=[str(send_res.get("error"))],
            )
            self.task_manager.update_task_status(task.task_id, TaskStatus.FAILED, result=err_result, increment_retry=True)
            return err_result

        # Poll status until IDLE or completed
        max_poll_time = 300
        poll_interval = 0.5
        elapsed = 0.0

        while elapsed < max_poll_time:
            time.sleep(poll_interval)
            elapsed += poll_interval

            status = self.worker_client.get_status(host, port)
            if status and status.current_task_id != task.task_id:
                # Task finished
                result = self.worker_client.get_result(host, port)
                if result and result.task_id == task.task_id:
                    # Evaluate result
                    eval_res = self.evaluator.evaluate_task_result(task, result)
                    if eval_res.verdict == EvaluationVerdict.PASS:
                        self.task_manager.update_task_status(task.task_id, TaskStatus.COMPLETED, result=result)
                    elif eval_res.verdict == EvaluationVerdict.RETRY:
                        self.task_manager.update_task_status(task.task_id, TaskStatus.RETRY, result=result, increment_retry=True)
                    else:
                        self.task_manager.update_task_status(task.task_id, TaskStatus.FAILED_PERMANENTLY, result=result, increment_retry=True)
                    return result

        timeout_result = TaskResult(
            task_id=task.task_id,
            agent_id=task.assigned_agent,
            status="FAILED",
            summary=f"Task {task.task_id} timed out while polling worker {task.assigned_agent}",
            errors=["Timeout"],
        )
        self.task_manager.update_task_status(task.task_id, TaskStatus.FAILED, result=timeout_result, increment_retry=True)
        return timeout_result

    def run_orchestration_loop(self, max_iterations: int = 50) -> bool:
        """Run tasks in dependency order until all are completed or permanently failed."""
        iterations = 0

        while iterations < max_iterations:
            iterations += 1
            ready_tasks = self.task_manager.get_next_ready_tasks()
            if not ready_tasks:
                break

            for task in ready_tasks:
                print(f"[*] Dispatching task [{task.task_id}] '{task.title}' to [{task.assigned_agent}]...")
                result = self.execute_single_task(task)
                print(f"    -> Result: {result.status} | {result.summary}")

        all_done = self.task_manager.is_all_completed()
        return all_done

File: crates/mag-manager/src/orchestrator.rs
//! Multi-Agent Orchestration coordinator in Rust.

use crate::evaluator::{EvaluationVerdict, ResultEvaluator};
use mag_agent::AgentDefinition;
use mag_common::{AgentRole, TaskRequest, TaskResult};
use mag_config::ProjectConfig;
use mag_git::GitManager;
use mag_scheduler::TaskScheduler;
use mag_storage::Storage;
use mag_task::Task;
use mag_worker::execute_task_for_agent;
use std::collections::HashMap;
use std::io::Write;
use std::path::{Path, PathBuf};
use thiserror::Error;

#[derive(Error, Debug)]
pub enum OrchestratorError {
    #[error("Storage error: {0}")]
    Storage(#[from] mag_storage::StorageError),
    #[error("Git error: {0}")]
    Git(#[from] mag_git::GitError),
    #[error("Task execution failed permanently: {0}")]
    Fatal(String),
}

pub struct Orchestrator {
    pub config: ProjectConfig,
    pub storage: Storage,
    pub git: GitManager,
    pub evaluator: ResultEvaluator,
    pub repo_path: PathBuf,
}

impl Orchestrator {
    pub fn new<P: AsRef<Path>>(repo_path: P, db_path: P) -> Result<Self, OrchestratorError> {
        let repo_buf = repo_path.as_ref().to_path_buf();
        let storage = Storage::new(db_path)?;
        let git = GitManager::new(&repo_buf);
        let config = ProjectConfig::default_project("mag-project");
        let evaluator = ResultEvaluator::new(config.manager.max_retries);

        Ok(Self {
            config,
            storage,
            git,
            evaluator,
            repo_path: repo_buf,
        })
    }

    pub fn extract_target_directory(&self, prompt: &str) -> PathBuf {
        for word in prompt.split_whitespace() {
            let trimmed = word.trim_matches(|c| c == '\'' || c == '"' || c == '`' || c == 'に' || c == 'へ');
            if trimmed.starts_with('/') || trimmed.starts_with("./") {
                return PathBuf::from(trimmed);
            }
        }
        self.repo_path.clone()
    }

    pub fn decompose_requirement(&self, prompt: &str, worker_count: Option<usize>) -> Result<Vec<Task>, OrchestratorError> {
        let existing = self.storage.list_tasks()?;
        let base_num = existing.len() + 1;

        let id1 = format!("TASK-{:03}", base_num);
        let id2 = format!("TASK-{:03}", base_num + 1);
        let id3 = format!("TASK-{:03}", base_num + 2);
        let id4 = format!("TASK-{:03}", base_num + 3);
        let id5 = format!("TASK-{:03}", base_num + 4);

        let t1 = Task::new(&id1, format!("Spec: {}", prompt), format!("Architecture notes for: {}", prompt), "agent-e", AgentRole::Researcher);
        let t2 = Task::new(&id2, format!("Implementation: {}", prompt), format!("Implement source for: {}", prompt), "agent-a", AgentRole::Developer)
            .with_dependencies(vec![id1.clone()]);
        let t3 = Task::new(&id3, format!("Testing: {}", prompt), format!("Run tests for: {}", prompt), "agent-b", AgentRole::Tester)
            .with_dependencies(vec![id2.clone()]);
        let t4 = Task::new(&id4, format!("Review: {}", prompt), format!("Review code for: {}", prompt), "agent-c", AgentRole::Reviewer)
            .with_dependencies(vec![id3.clone()]);
        let t5 = Task::new(&id5, format!("Security: {}", prompt), format!("Security audit for: {}", prompt), "agent-d", AgentRole::Security)
            .with_dependencies(vec![id4.clone()]);

        let mut tasks = vec![t1, t2, t3, t4, t5];
        if let Some(count) = worker_count {
            TaskScheduler::assign_collaborative_workers(&mut tasks, count);
        }

        for t in &tasks {
            self.storage.save_task(t)?;
        }

        Ok(tasks)
    }

    pub fn decompose_requirement_with_agents(
        &self,
        prompt: &str,
        active_agents: &[String],
    ) -> Result<Vec<Task>, OrchestratorError> {
        let existing = self.storage.list_tasks()?;
        let base_num = existing.len() + 1;

        let id1 = format!("TASK-{:03}", base_num);
        let id2 = format!("TASK-{:03}", base_num + 1);
        let id3 = format!("TASK-{:03}", base_num + 2);
        let id4 = format!("TASK-{:03}", base_num + 3);
        let id5 = format!("TASK-{:03}", base_num + 4);

        let mut t1 = Task::new(&id1, format!("Spec: {}", prompt), format!("Architecture notes for: {}", prompt), "agent-e", AgentRole::Researcher);
        let mut t2 = Task::new(&id2, format!("Implementation: {}", prompt), format!("Implement source for: {}", prompt), "agent-a", AgentRole::Developer)
            .with_dependencies(vec![id1.clone()]);
        let mut t3 = Task::new(&id3, format!("Testing: {}", prompt), format!("Run tests for: {}", prompt), "agent-b", AgentRole::Tester)
            .with_dependencies(vec![id2.clone()]);
        let mut t4 = Task::new(&id4, format!("Review: {}", prompt), format!("Review code for: {}", prompt), "agent-c", AgentRole::Reviewer)
            .with_dependencies(vec![id3.clone()]);
        let mut t5 = Task::new(&id5, format!("Security: {}", prompt), format!("Security audit for: {}", prompt), "agent-d", AgentRole::Security)
            .with_dependencies(vec![id4.clone()]);

        if !active_agents.is_empty() {
            let n = active_agents.len();
            t1.assigned_agent = active_agents[0 % n].clone();
            t2.assigned_agent = active_agents[1 % n].clone();
            t3.assigned_agent = active_agents[2 % n].clone();
            t4.assigned_agent = active_agents[3 % n].clone();
            t5.assigned_agent = active_agents[4 % n].clone();
        }

        let tasks = vec![t1, t2, t3, t4, t5];
        for t in &tasks {
            self.storage.save_task(t)?;
        }

        Ok(tasks)
    }

    pub fn init_task_md(&self, prompt: &str, tasks: &[Task]) -> Result<(), std::io::Error> {
        let mut md = String::new();
        md.push_str("# Multi-Agent Task Execution Log (`task.md`)\n\n");
        md.push_str(&format!("**Requirement / Prompt:** `{}`\n\n", prompt));
        md.push_str("## 📋 Execution Plan (Task DAG)\n\n");
        md.push_str("| Task ID | Role | Assigned Agent | Status | Dependencies |\n");
        md.push_str("|---|---|---|---|---|\n");
        for t in tasks {
            let deps = if t.dependencies.is_empty() { "root".into() } else { t.dependencies.join(", ") };
            md.push_str(&format!("| **{}** | `{}` | `{}` | `{}` | `{}` |\n", t.id, t.role, t.assigned_agent, t.status, deps));
        }
        md.push_str("\n---\n\n## 🔄 Real-Time Execution & Evaluation History\n\n");
        let path = self.repo_path.join("task.md");
        std::fs::write(path, md)
    }

    pub fn append_task_log(&self, task: &Task, res: &TaskResult) -> Result<(), std::io::Error> {
        let mut md = String::new();
        md.push_str(&format!("### 🔹 [{}] {}\n\n", task.id, task.title));
        md.push_str(&format!("- **Assigned Agent:** `{}` ({})\n", task.assigned_agent, task.role));
        md.push_str(&format!("- **Status:** `{}`\n", task.status));
        md.push_str(&format!("- **Execution Verdict:** `{}`\n", res.status));
        md.push_str(&format!("- **Summary:** {}\n", res.summary));
        if !res.files_changed.is_empty() {
            md.push_str(&format!("- **Files Modified:** `{}`\n", res.files_changed.join(", ")));
        }
        if task.retry_count > 0 {
            md.push_str(&format!("- **Self-Repair Retries:** {}/{}\n", task.retry_count, task.max_retries));
        }
        if !res.errors.is_empty() {
            md.push_str("\n**Errors / Diagnostic Logs:**\n```text\n");
            for err in &res.errors {
                md.push_str(err);
                md.push('\n');
            }
            md.push_str("```\n");
        }
        md.push('\n');

        let path = self.repo_path.join("task.md");
        let mut file = std::fs::OpenOptions::new().create(true).append(true).open(path)?;
        file.write_all(md.as_bytes())
    }

    pub fn finalize_task_md(&self, success: bool) -> Result<(), std::io::Error> {
        let mut md = String::new();
        md.push_str("---\n\n## 📊 Final Workflow Summary\n\n");
        if success {
            md.push_str("✅ **Status:** `ALL TASKS COMPLETED & VERIFIED SUCCESSFULLY`\n\n");
            md.push_str("- **Manager Evaluation:** `APPROVED`\n");
            md.push_str("- **Branch Status:** Merged into `main` branch\n");
        } else {
            md.push_str("⚠️ **Status:** `WORKFLOW FINISHED WITH ERRORS`\n\n");
        }
        let path = self.repo_path.join("task.md");
        let mut file = std::fs::OpenOptions::new().create(true).append(true).open(path)?;
        file.write_all(md.as_bytes())
    }

    pub fn execute_task_locally(&self, task: &mut Task, target_dir: Option<&Path>) -> Result<TaskResult, OrchestratorError> {
        let agent = AgentDefinition::default_for_role(task.role);
        let effective_repo = target_dir.unwrap_or(&self.repo_path);
        let repo_str = effective_repo.to_string_lossy().to_string();

        let mut metadata: HashMap<String, serde_json::Value> = HashMap::new();

        match task.role {
            AgentRole::Researcher => {
                let doc = format!(
                    "# Specification Document\n\n- Task: {}\n- Target: {}\n- Summary: Auto-generated spec by Agent-E.\n",
                    task.title, repo_str
                );
                metadata.insert("doc_content".into(), serde_json::Value::String(doc));
            }
            AgentRole::Developer => {
                let pkg_name = effective_repo
                    .file_name()
                    .and_then(|s| s.to_str())
                    .unwrap_or("my_app")
                    .replace('-', "_");

                let cargo_toml = format!(
                    "[package]\nname = \"{}\"\nversion = \"0.1.0\"\nedition = \"2021\"\n\n[dependencies]\n",
                    pkg_name
                );

                let main_rs = format!(
                    "//! Auto-generated by Multi-Agent Development Orchestrator (`mag`)\n\nfn main() {{\n    println!(\"Hello from mag multi-agent orchestrator!\");\n    let val = calculate_sum(10, 20);\n    println!(\"Sample calculation result: {{}}\", val);\n}}\n\npub fn calculate_sum(a: i32, b: i32) -> i32 {{\n    a + b\n}}\n\n#[cfg(test)]\nmod tests {{\n    use super::*;\n\n    #[test]\n    fn test_sum() {{\n        assert_eq!(calculate_sum(10, 20), 30);\n    }}\n}}\n"
                );

                metadata.insert(
                    "files".into(),
                    serde_json::json!({
                        "Cargo.toml": cargo_toml,
                        "src/main.rs": main_rs
                    }),
                );
                metadata.insert("command".into(), serde_json::Value::String("cargo check || true".into()));
            }
            AgentRole::Tester => {
                metadata.insert(
                    "test_command".into(),
                    serde_json::Value::String("cargo test || cargo check || python3 -m unittest discover || true".into()),
                );
            }
            _ => {}
        }

        let task_req = TaskRequest {
            task_id: task.id.clone(),
            r#type: task.role.to_string(),
            title: task.title.clone(),
            description: task.description.clone(),
            repository: repo_str,
            branch: format!("{}/{}", task.assigned_agent, task.id.to_lowercase()),
            timeout_seconds: 300,
            context_files: vec![],
            metadata,
        };

        let result = execute_task_for_agent(&agent, &task_req);
        let (verdict, next_status) = self.evaluator.evaluate(task, &result);

        task.status = next_status;
        task.result = Some(result.clone());
        if matches!(verdict, EvaluationVerdict::Retry { .. }) {
            task.retry_count += 1;
        }

        self.storage.save_task(task)?;
        let _ = self.append_task_log(task, &result);
        Ok(result)
    }

    pub fn run_orchestration_loop(&self, target_dir: Option<&Path>, max_iterations: usize) -> Result<bool, OrchestratorError> {
        let mut iterations = 0;

        while iterations < max_iterations {
            iterations += 1;
            let all_tasks = self.storage.list_tasks()?;
            let ready_tasks = TaskScheduler::get_ready_tasks(&all_tasks);

            if ready_tasks.is_empty() {
                break;
            }

            for mut task in ready_tasks {
                println!("[*] Dispatching task [{}] '{}' to [{}]...", task.id, task.title, task.assigned_agent);
                let res = self.execute_task_locally(&mut task, target_dir)?;
                println!("    -> Result: {} | {}", res.status, res.summary);
            }
        }

        let final_tasks = self.storage.list_tasks()?;
        let all_passed = TaskScheduler::is_all_completed(&final_tasks);
        let _ = self.finalize_task_md(all_passed);

        if all_passed {
            let eff_dir = target_dir.unwrap_or(&self.repo_path);
            let git_mgr = GitManager::new(eff_dir);
            if !git_mgr.is_repo() {
                let _ = git_mgr.init_repo();
            }
        }

        Ok(all_passed)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[test]
    fn test_orchestrator_lifecycle() {
        let dir = tempdir().unwrap();
        let db_path = dir.path().join("test.sqlite");

        let orch = Orchestrator::new(dir.path(), &db_path).unwrap();
        let tasks = orch.decompose_requirement("Create hello library in Rust", None).unwrap();
        assert_eq!(tasks.len(), 5);

        let _ = orch.init_task_md("Create hello library in Rust", &tasks);
        let success = orch.run_orchestration_loop(Some(dir.path()), 10).unwrap();
        assert!(success);
        assert!(dir.path().join("Cargo.toml").exists());
        assert!(dir.path().join("src/main.rs").exists());
        assert!(dir.path().join("task.md").exists());
    }

    #[test]
    fn test_orchestrator_two_workers() {
        let dir = tempdir().unwrap();
        let db_path = dir.path().join("test.sqlite");

        let orch = Orchestrator::new(dir.path(), &db_path).unwrap();
        let tasks = orch.decompose_requirement("Build math utility in Rust", Some(2)).unwrap();
        assert_eq!(tasks[0].assigned_agent, "agent-1");
        assert_eq!(tasks[1].assigned_agent, "agent-2");
        assert_eq!(tasks[2].assigned_agent, "agent-1");
    }

    #[test]
    fn test_orchestrator_active_agents_dispatch() {
        let dir = tempdir().unwrap();
        let db_path = dir.path().join("test.sqlite");

        let orch = Orchestrator::new(dir.path(), &db_path).unwrap();
        let active = vec!["agent-a".to_string(), "agent-b".to_string()];
        let tasks = orch.decompose_requirement_with_agents("Test active dispatch", &active).unwrap();
        assert_eq!(tasks[0].assigned_agent, "agent-a");
        assert_eq!(tasks[1].assigned_agent, "agent-b");
        assert_eq!(tasks[2].assigned_agent, "agent-a");
    }
}
